rustdoc: Allow multiple references to a single footnote

a4lg opened the pull request:
Multiple references to a single footnote is a part of GitHub Flavored Markdown syntax (although not explicitly documented as well as regular footnotes, it is implemented in GitHub's fork of CommonMark) and not prohibited by rustdoc.

cf. <https://github.com/github/cmark-gfm/blob/587a12bb54d95ac37241377e6ddc93ea0e45439b/test/extensions.txt#L762-L780>

However, using it makes multiple `sup` elements with the same `id` attribute, which is invalid per the HTML specification.

Still, not only this is a valid GitHub Flavored Markdown syntax, this is helpful on certain cases and actually tested (accidentally) in `tests/rustdoc/footnote-reference-in-footnote-def.rs`.

This commit keeps track of the number of references per footnote and gives unique ID to each reference.
It also emits *all* back links from a footnote to its references as "↩" (return symbol) plus a numeric list in superscript.

As a known limitation, it assumes that all references to a footnote are rendered (this is not always true if a dangling footnote has one or more references but considered a reasonable compromise).

Also note that, this commit is designed so that no HTML changes will occur unless multiple references to a single footnote is actually used.

<!-- homu-ignore:start -->

# Background

A failure is detected on the CI process of rust-lang/rust#140389, which adopted `stdarch` submodule with PR rust-lang/stdarch#1779.

As you see in the screenshot of that stdarch PR, it uses multiple references to a single footnote to simplify showing various platform/version-specific notes (this is far more important than x86 because there are no architectural, fine-grained feature detection methods on RISC-V and hence feature detection is highly platform/version-specific).

And I thought this kind of references are allowed because:

1.  `rustdoc` does not reject such links and
2.  One of the tests `tests/rustdoc/footnote-reference-in-footnote-def.rs` contains multiple references to a single footnote `[^a]`.

...until I encounter a linkchecker failure (ran on CI).

# Proposal

Of course, rejecting such references might be an option but this PR attempts to resolve the issue by explicitly allowing multiple references to a single footnote by:

1.  Generating unique `id` attribute per a reference to a footnote and
2.  Emitting *all* back links from a footnote as a return symbol plus a numeric list in superscript.

Note that, this PR is designed so that no HTML changes will occur unless multiple references to a single footnote is *actually used*.

# Known Limitation

To simplify the implementation, it just keeps track of the number of references per footnote and *assumes* that all references to footnotes are rendered.  This is usually true but may not be *always* true if a reference is inside a footnote and that footnote is dangling (in this case, broken links (that won't cause any action when clicked/touched) will be generated but otherwise fine).

I also didn't use atomics but was it necessary?

# Screenshot

This is a screenshot of this PR plus rust-lang/stdarch#1779, showing how back links are rendered.
You can see the differences in the footnotes 1, 2 and 5.

![Screenshot of stdarch PR 1779 with this PR (footnote part)](https://github.com/user-attachments/assets/06795963-36d4-493e-ba5c-1756f6240ade)

# Design Considerations / Options

I just used 1-origin numbers for back links from a footnote but if this is confusing, using alphabet-based list "a", "b"..."z", "aa", "ab"... might be an option like in Wikipedia.  The reason I didn't do this (in the first proposal) is because back links are visually distinct than Wikipedia and seems easy to make distinction between regular references to footnotes due to the return symbol "↩" (Wikipedia: caret "^" is used).

# History

## Version 1 (2025-04-29)

The initial proposal.

## Version 2 and 3 (2025-04-30)

Excluding the rebase, they only change the commit message (mainly grammar fixes)
and the code is unchanged from the version 1.

## Version 4 (2025-05-06)

No changes in the source code.

*   Commit Message: Added minor adjustments
*   Commit Message: Added references to GitHub Flavored Markdown syntax and GitHub's fork of CommonMark
*   Commit Message: Clarified that multiple references in `tests/rustdoc/footnote-reference-in-footnote-def.rs` are (most likely) not to test such references themselves.

## Version 5 (2025-08-14)

*   Reflect the feedback by @fmease.
*   Commit Message: "cause" → "make"

<!-- homu-ignore:end -->

notriddle commented:
I'm satisfied with the known limitation on nested, hidden footnotes.

Since this is a UI-visible change, I'll cc the frontend team about it.

@rfcbot poll appearance

GuillaumeGomez commented:
It's really cool, nicely done!

a4lg commented:
With all due respect, if time constraints allow, I would greatly appreciate it if this PR could be merged within the version 1.88 cycle (before beta branching occurs).  That would allow re-applying platform-specific guide documentation (RISC-V) on `stdarch` in the version 1.89 cycle.
Thank you for your consideration.

a4lg commented:
The syntax of using multiple references to a single footnote is compatible to [GitHub Flavored Markdown](https://github.com/github/cmark-gfm/blob/587a12bb54d95ac37241377e6ddc93ea0e45439b/test/extensions.txt#L762-L780)
but with a slight design differences in both:

*   Footnote reference
    *   GitHub: brackets are used like `[1]`
    *   rustdoc (current implementation): raw superscript is used like `1`) and
*   Back link list
    *   GitHub: every back links have the "return" symbol
    *   rustdoc (this PR): only the first back link has the "return" symbol).

For instance, GitHub Flavored Markdown source code:

```
This is an example [^a] of multiple references [^a] to a single footnote [^a].

[^a]: This is the footnote.
```

is rendered as below:

This is an example [^a] of multiple references [^a] to a single footnote [^a].

[^a]: This is the footnote.

a4lg commented:
If we apply GitHub's design in back links (for rustdoc source code which generated the image below, see [my experimental branch](https://github.com/a4lg/rust/tree/experimental/rustdoc-multi-footnote-refs/github-like-backlinks), it will look like this:

![image](https://github.com/user-attachments/assets/6a79a9b6-763a-4c65-96bc-9beda4080807)

In my opinion, it looks a bit cluttered.

notriddle commented:
I think it’s not a good idea to worry about how bad it looks when there are nine references to one footnote. Even if it looked better, the user still needs to pick one, and it’s not clear how to tell which one you want. So, if you have eight refs, you have bigger problems than appearance.

On the other hand, the GFM approach makes the link bigger, so it’s easier to click or tap.

a4lg commented:
Excuse me.
Did I forget something to do on my side?  If so, I appreciate it if you could let me know.

To note about the Markdown syntax, this PR does not extend the existing syntax (only changes the behavior where rustdoc allowed but with invalid HTML output, which is fixed by this PR).

bors commented:
:umbrella: The latest upstream changes (presumably #143363) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

a4lg commented:
Resolved merge conflict with #134006.

@camelid @jsha Feedback is appreciated.

fmease commented on src/librustdoc/html/markdown/footnotes.rs:
```diff
@@ -33,21 +35,28 @@ impl<'a, I: Iterator<Item = SpannedEvent<'a>>> Footnotes<'a, I> {
-        let (_, id) = self.get_entry(reference);
+        let (_, id, num_refs) = self.get_entry(reference);
+        *num_refs += 1;
+        let fnref_suffix = {
+            let num_refs = *num_refs;
+            if num_refs <= 1 { "".to_owned() } else { format!("-{num_refs}") }
+        };
```
```suggestion
        let fnref_suffix =
            if *num_refs <= 1 { "".to_owned() } else { format!("-{num_refs}") };
```

a4lg replied:
Oh, I used `{x}` (when `x` is a reference) in my programs and I somehow ignored [implementations for references](https://doc.rust-lang.org/1.89.0/std/fmt/trait.Display.html#impl-Display-for-%26mut+T).  Will change.

fmease commented:
> On the other hand, the GFM approach makes the link bigger, so it’s easier to click or tap.

I was about to say we should go with GH's rendering but then I realized that consecutive footnote *references* have basically the same style: <nobr><sup>1</sup> <sup>2</sup> <sup>3</sup></nobr> (same small size, not a lot of spacing) whereas on GH, they're rendered with extra square brackets: <nobr><sup>[1]</sup> <sup>[2]</sup> <sup>[3]</sup></nobr> (of which I'm not sure if they meaningfully increase the clickable area). And so far, nobody seems have complained about that so the current approach seems acceptable to me for now.

fmease commented:
Since camelid and jsha haven't replied so far and since this GUI change seems small and trivial enough (also, no FCP, only a poll), I'm willing to give an "official +1" for this despite not being on T-rustdoc-frontend if it means this PR can be unblocked.

a4lg committed:
rustdoc: Allow multiple references to a single footnote

Multiple references to a single footnote is a part of GitHub Flavored
Markdown syntax (although not explicitly documented as well as regular
footnotes, it is implemented in GitHub's fork of CommonMark) and not
prohibited by rustdoc.

cf. <https://github.com/github/cmark-gfm/blob/587a12bb54d95ac37241377e6ddc93ea0e45439b/test/extensions.txt#L762-L780>

However, using it makes multiple "sup" elements with the same "id"
attribute, which is invalid per the HTML specification.

Still, not only this is a valid GitHub Flavored Markdown syntax, this is
helpful on certain cases and actually tested (accidentally) in
tests/rustdoc/footnote-reference-in-footnote-def.rs.

This commit keeps track of the number of references per footnote and gives
unique ID to each reference.  It also emits *all* back links from a footnote
to its references as "↩" (return symbol) plus a numeric list in superscript.

As a known limitation, it assumes that all references to a footnote are
rendered (this is not always true if a dangling footnote has one or more
references but considered a reasonable compromise).

Also note that, this commit is designed so that no HTML changes will occur
unless multiple references to a single footnote is actually used.

GuillaumeGomez commented:
Then let's go! Thanks everyone!

@bors r=fmease,GuillaumeGomez rollup
